### PR TITLE
feat: more flexible URI template util

### PR DESCRIPTION
**Pull request by dtbuchholz**
## Summary

The current setup with `URITemplate.sol` is rather rigid. It only allows for a single `{id}` to exist when templating, and the Rigs implementation ended up using a more flexible approach. These changes reflect some initial thoughts toward how this can be implemented and demo out 'single' vs 'multiple' methods (for gas savings purposes) and a lib vs contract impl. TBD which is ideal.

## Details

To demonstrate the different approaches, `URITemplate` and `URITemplateLib` have been created. The first is a contract (the way this util is currently designed), and the second is a library. The rationale behind the lib wa because I was messing around with multiple methods that alter the same storage variable but with different behavior. 

One "set URI template" method only allows for a single `{id}`, and the other can handle multiple `{id}` substrings. Keeping both methods might be worthwhile for gas savings purposes. I wasn't sure which was the better approach, so both are included in this initial pass.
- `TestURITemplate` inherits from a `URITemplate` contract
- `TestURITemplateLib` uses a `URITemplateLib` library
- Both test contracts have methods to `setURITemplateSingle` or `setURITemplateMultiple` (re: single or multiple `{id}`s)
- Realistically, all that's probably needed here is the contract impl with only `_setURITemplateMultiple` and `_getURITemplateMultiple`, but I figured I'd try the other options out

## How it was tested

Tests have been added in the `test` directory, with gas comparisons shown below:
```
·······················|··························|·············|·············|·············|···············|··············
|  Contract            ·  Method                  ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplate     ·  mint                    ·      56296  ·      90496  ·      83656  ·           10  ·          -  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplate     ·  setURITemplateMultiple  ·      80169  ·     162348  ·     115943  ·            3  ·          -  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplate     ·  setURITemplateSingle    ·      52044  ·     101384  ·      84857  ·            3  ·          -  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplateLib  ·  mint                    ·          -  ·          -  ·      90496  ·            6  ·          -  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplateLib  ·  setURITemplateMultiple  ·      80181  ·     162360  ·     115955  ·            3  ·          -  │
·······················|··························|·············|·············|·············|···············|··············
|  TestURITemplateLib  ·  setURITemplateSingle    ·          -  ·          -  ·     112182  ·            1  ·          -  │
```

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by dtbuchholz** [comment]
@asutula lmk if this is how you were thinking about things. I ended up creating both a lib and contract setup to test things out. These aren't necessarily gas optimized, but I wanted to see which felt more natural between the lib/contract. 

There are also two different methods for setting the token URI. My initial thought was that the implementer may only need a single `{id}`, so they'd want to use a more gas efficient `_setURITemplateSingle` method. The `_setURITemplateMultiple` is the more flexible one, with 1+ allowed `{id}`s.

File: test/unit/utils/URITemplate.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    const minter = accounts[1];
+    const tx = await contract.connect(minter).mint();
+    const receipt = await tx.wait();
+    const [event] = receipt.events ?? [];
+    const tokenId = event.args?.tokenId;
+    // Check the token URI is the incorrect format
+    expect(await contract.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/bar={id}/foo`
+    );
+  });
```
Comment (asutula) [inline test/unit/utils/URITemplate.ts:76]: Is there any reasonable way to make `contract.setURITemplateSingle` error/revert in this case? Seems strange to let it all succeed and end up with results you didn't expect.

File: test/unit/utils/URITemplate.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    await lib.setURITemplateMultiple("https://foo.xyz/{id}/asdf");
+    // Check the token URI is the correct format
+    expect(await lib.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/asdf`
+    );
+  });
+
+  it("Should fail to create URI with multiple 'id' using the 'single' method", async function () {
+    // Set token URI
+    await contract.setURITemplateSingle("https://foo.xyz/{id}/bar={id}/foo");
```
Comment (asutula) [inline test/unit/utils/URITemplate.ts:156]: Same comment as about about if this can just fail.

**Review by asutula — COMMENTED** [review]
Nice, looking really good. Left a couple comments/questions. My initial reaction is that the contract approach feels better just because it takes care of the storage and requires less boilerplate to use.

**Comment by asutula** [comment]
If you think the library approach is appealing to some people, we could provide both and have the contract use the library.

File: test/unit/utils/URITemplate.ts (hunk 1 above)
Comment (dtbuchholz) [reply test/unit/utils/URITemplate.ts:76]: @asutula totally, I'll add this! I initially didn't want to over engineer or write all of the tests until I was certain which approach I was taking, which was short sighted.

File: test/unit/utils/URITemplate.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    const minter = accounts[1];
+    const tx = await lib.connect(minter).mint();
+    const receipt = await tx.wait();
+    const [event] = receipt.events ?? [];
+    const tokenId = event.args?.tokenId;
+    expect(await lib.tokenURI(tokenId)).to.equal(
+      `https://foo.xyz/${tokenId}/bar`
+    );
+    // Set the URI again, with a longer string and another 'id' substring
+    await lib.setURITemplateMultiple("https://foo.xyz/{id}/asdf");
```
[the remark this answers is not in the document]
Comment (dtbuchholz) [reply test/unit/utils/URITemplate.ts:147]: Good catch! (fwiw, I hadn't fully cleaned up the tests quite yet)

**Comment by dtbuchholz** [comment]
Agreed re: contract...abstracts the impl details. Probably no need for a lib, either.

**Comment by sanderpick** [comment]
Hey guys, just stubbed on this via email. To add some context, the rigs contract had avoid parsing the template string on-chain because it was so huge... iirc, i was having trouble estimating gas for so much string parsing work. So, I went with an approach that simply parses the string locally via https://github.com/tablelandnetwork/rigs/blob/main/ethereum/helpers/uris.ts, and then passes a list of "template parts" to the contract (https://github.com/tablelandnetwork/rigs/blob/main/ethereum/contracts/utils/URITemplate.sol#L20). I think that is the correct way to handle this for a couple reasons:
- the on-chain string parsing is gassy
- the string parsing utils (`SStrings`) unnecessarily adds to contact size
- it doesn't work / is finicky with very large strings, like many Tableland queries

**Comment by dtbuchholz** [comment]
Gotcha...I'd agree with that, due to the amount of gas associated with the on-chain splicing/splitting/etc. I was trying to make it work but doesn't seem like it'll ever be efficient enough.

I'll just make it simple and update this to the Rigs flavor.

**Pull request merged**